pkg/apis/kudo/v1alpha1/instance_types.go

```diff
@@ -16,65 +16,372 @@ limitations under the License.
+func (s ExecutionStatus) IsRunning() bool {
+	return s == ExecutionInProgress || s == ExecutionPending || s == ErrorStatus
+}
+
+// GetPlanInProgress returns plan status of currently active plan or nil if no plan is running
+func (i *Instance) GetPlanInProgress() *PlanStatus {
+	for _, p := range i.Status.PlanStatus {
+		if p.Status.IsRunning() {
+			return &p
+		}
+	}
+	return nil
+}
+
+// NoPlanEverExecuted returns true is this is new instance for which we never executed any plan
+func (i *Instance) NoPlanEverExecuted() bool {
+	for _, p := range i.Status.PlanStatus {
+		if p.Status != ExecutionNeverRun {
+			return false
+		}
+	}
+	return true
+}
+
+// EnsurePlanStatusInitialized initializes plan status for all plans this instance supports
+// it does not trigger run of any plan
+// it either initializes everything for a fresh instance without any status or tries to adjust status after OV was updated
+func (i *Instance) EnsurePlanStatusInitialized(ov *OperatorVersion) {
+	if i.Status.PlanStatus == nil {
+		i.Status.PlanStatus = make(map[string]PlanStatus)
+	}
+
+	for planName, plan := range ov.Spec.Plans {
+		planStatus := &PlanStatus{
+			Name:   planName,
+			Status: ExecutionNeverRun,
+			Phases: make([]PhaseStatus, 0),
+		}
+
+		existingPlanStatus, planExists := i.Status.PlanStatus[planName]
+		if planExists {
+			planStatus.Status = existingPlanStatus.Status
+		}
+		for _, phase := range plan.Phases {
+			phaseStatus := &PhaseStatus{
+				Name:   phase.Name,
+				Status: ExecutionNeverRun,
+				Steps:  make([]StepStatus, 0),
+			}
+			existingPhaseStatus, phaseExists := PhaseStatus{}, false
+			if planExists {
+				for _, oldPhase := range existingPlanStatus.Phases {
+					if phase.Name == oldPhase.Name {
+						existingPhaseStatus = oldPhase
+						phaseExists = true
+						phaseStatus.Status = existingPhaseStatus.Status
+					}
+				}
+			}
+			for _, step := range phase.Steps {
+				stepStatus := StepStatus{
+					Name:   step.Name,
+					Status: ExecutionNeverRun,
+				}
+				if phaseExists {
+					for _, oldStep := range existingPhaseStatus.Steps {
+						if step.Name == oldStep.Name {
+							stepStatus.Status = oldStep.Status
+						}
+					}
+				}
+				phaseStatus.Steps = append(phaseStatus.Steps, stepStatus)
+			}
+			planStatus.Phases = append(planStatus.Phases, *phaseStatus)
+		}
+		i.Status.PlanStatus[planName] = *planStatus
+	}
+}
+
+// StartPlanExecution mark plan as to be executed
+func (i *Instance) StartPlanExecution(planName string, ov *OperatorVersion) error {
+	if i.NoPlanEverExecuted() || isUpgradePlan(planName) {
+		i.EnsurePlanStatusInitialized(ov)
+	}
+
+	// update status of the instance to reflect the newly starting plan
+	notFound := true
+	for n1, v := range i.Status.PlanStatus {
+		if v.Name == planName {
+			// update plan status
+			notFound = false
+			planStatus := i.Status.PlanStatus[n1]
+			planStatus.Status = ExecutionPending
+			for i2, p := range v.Phases {
```

Comment:
```suggestion
			for j, p := range v.Phases {
```
j and k are normally acceptable index vars like i.   could we use j and k instead of i2 and i3?